### PR TITLE
fix(learning): forget_facet must not log a drop that never happened

### LINKED ISSUE #6108: `learning.forget_facet` reports success — and logs a state change — for a facet that does not exist
## Summary

`learning.forget_facet` reports success — and logs a state change — for a facet that does not exist

Found in the e2e coverage wave, independently verified by a second engineer, and the fix direction
has now been **decided by the maintainer** — see Solution.

- **Namespace / method:** `openhuman.learning_forget_facet`
- **Where:** `src/openhuman/agent/learning/schemas_part_02.rs:228-279`
- **Verified at:** `openhuman` `main` @ `d1ed68f9b`

## Problem

**Expected**

the same contract as its three sibling mutators. `update_facet`,
`pin_facet` and `unpin_facet` all refuse an unknown key with
`facet not found: <class>/<key>`.

**Actual**

`forget_facet` reads the row, and when it is absent takes the
`else` branch, sets `facet_json = Value::Null`, and returns `Ok` — with the
log line
```
learning.forget_facet: key=<class>/<key> state=dropped user_state=forgotten
```
built unconditionally, before the branch is known. So for a typo'd key the
handler answers success and writes a log claiming it dropped a row it never
touched. Nothing was written; nothing says so.

**Steps to reproduce**

`openhuman.learning_forget_facet` with
`{"class": "style", "key": "e2e_learning_key_that_was_never_observed"}`
→ `{"facet": null}`, HTTP 200, no JSON-RPC error, and the log line above.
The same key on `pin_facet` → `facet not found: style/e2e_learning_key_…`.

### Independent verification

**Repro run:** `cargo test --test raw_coverage_all --features "$(bash scripts/ci/product-features.sh)" -- --nocapture zzz_w2_verify_probe`
(scratch module, deleted after the run) driving the real JSON-RPC surface with memory seams +
`set_modules_policy` published.

**Observed**

```
forget_facet(style/w2_never_observed):
  {"result":{"logs":["learning.forget_facet: key=style/w2_never_observed state=dropped user_state=forgotten"],
             "result":{"facet":null}}}
pin_facet   : {"error":{"code":-32000,"message":"facet not found: style/w2_never_observed"}}
unpin_facet : {"error":{"code":-32000,"message":"facet not found: style/w2_never_observed"}}
update_facet: {"error":{"code":-32000,"message":"facet not found: style/w2_never_observed"}}
```
So both halves reproduce verbatim: HTTP 200, no JSON-RPC error, and a log line asserting
`state=dropped user_state=forgotten` about a row that was never read back, while all three
siblings refuse the identical key.

**Diagnosis assessment:** the *mechanism* is exactly right. But the report frames the leniency
as an oversight, and the evidence says it is a choice: the **agent-tool** implementation of the
same operation, `LearningForgetFacetTool::execute`
(`src/openhuman/agent/learning/tools.rs:387-412`), independently does the same thing —
`None => serde_json::Value::Null`, `ToolResult::success` — while the pin/unpin tool path
(`set_pin`, :271-294) errors with `facet not found: {fk}` just like the RPC. Two separate
implementations agree that `forget` is lenient and `pin`/`unpin`/`update` are strict. That
reads as an intentional idempotent-delete contract, not a slip.

## Solution

**DECIDED: make it idempotent.** Return `{"facet": null}` for an absent key.

Rationale: the agent tool at `tools.rs:387-412` already behaves idempotently, so this removes a divergence rather than creating one; delete-is-idempotent is the conventional RPC semantic; and a `forget` that errors on an absent key **leaks existence information** — the caller learns whether the facet was there, which is the wrong direction for a privacy operation.

**Apply to BOTH the RPC handler and `tools.rs:387-412`** or they diverge again.

Independent of the decision: move the `log` construction inside the `if let Some(...)` arm so the "dropped" claim is only made when a row was actually written.

**Blast radius:** the RPC handler and the agent tool are the two call sites; a change to the
RPC's error contract is visible to any frontend forget button and to
`learning_facet_controllers_refuse_bad_input_and_absent_keys`
(`tests/raw_coverage/learning_facets_e2e.rs`), which currently pins today's behaviour with a
comment pointing at this file. Note `reset_cache` → `cache::reset_non_pinned` is a *different*
path and is unaffected.

## Acceptance criteria

- [ ] **Repro gone** — the reproduction above no longer produces the reported behaviour.
- [ ] **Regression safety** — a test pins the decided behaviour so it cannot drift back.
- [ ] **Diff coverage ≥ 80%** — the fix PR meets the changed-lines coverage gate.
- [ ] **Blast radius checked** — every consumer named above still behaves correctly.

## Related

From the e2e coverage wave (#6053, #6054, #6055, #6058, #6059, #6060).

### PR DESCRIPTION
**Pull request by M3gA-Mind**
## Summary

- `learning.forget_facet` no longer logs `state=dropped user_state=forgotten` for a key that does not exist.
- The RPC contract is **unchanged**: an absent key still answers `{"facet": null}` with no error, per the decision on #6108.
- The log line moves behind `forget_facet_log(full_key, dropped)`, so the claim is made only when a row was actually written — and is unit-testable without a cache.

## Problem

`handle_forget_facet` read the row, branched on whether it existed, and then built its log line **unconditionally, after the branch and outside it**. For a typo'd key the handler answered success and wrote

```
learning.forget_facet: key=<class>/<key> state=dropped user_state=forgotten
```

for a row it never touched. Nothing was written; nothing said so.

Note this **inverts the original report's expectation.** The report wanted `forget_facet` to refuse an unknown key, matching its three sibling mutators (`update`/`pin`/`unpin`, which all return `facet not found: <class>/<key>`). The maintainer decided the opposite and the decision is implemented as written: forgetting stays idempotent because the agent tool already behaves that way, delete-is-idempotent is the conventional RPC semantic, and erroring **leaks existence information** — the caller learns whether the facet was there, which is the wrong direction for a privacy operation.

So the return value is deliberately unchanged, and the log is the only thing that can distinguish the two outcomes. It was saying the wrong one.

## Solution

- `forget_facet_log(full_key, dropped) -> Vec<String>` returns the "dropped" line only when a row was written, and `… not present — no change` otherwise.
- The `if let Some` arm now yields `(facet_json, dropped)` so the log is chosen from the branch that actually ran.
- **Checked, no change needed:** the decision said to apply this to both surfaces "or they diverge again". `tools.rs`'s `execute` already returns `Value::Null` for an absent key and emits no such log, so the two do not diverge. I verified rather than making a no-op edit.

## Submission Checklist

- [x] Tests added or updated (happy path + at least one failure / edge case) per Testing Strategy — two unit tests in `schemas_tests.rs`: the drop case still records the state change, the absent case must not.
- [x] **Diff coverage ≥ 80%** — every changed line is exercised by the two new tests; ran `cargo test --lib --features "$(bash scripts/ci/product-features.sh)"` locally. N/A for Vitest: no frontend code changed.
- [x] Coverage matrix updated — `N/A: behaviour-only change` to a log line; no feature row added, removed or renamed.
- [x] All affected feature IDs from the matrix are listed under `## Related` — `N/A: no matrix feature rows affected.`
- [x] No new external network dependencies introduced — no network code touched.
- [x] Manual smoke checklist updated if this touches release-cut surfaces — `N/A: not a release-cut surface; a debug log line on one RPC.`
- [x] Linked issue closed via `Closes #NNN` in the `## Related` section

## Impact

Log output only. No RPC contract change, no wire change, no frontend impact. The idempotent `{"facet": null}` response for an absent key is preserved deliberately.

**Revert-check.** Fault injected: `forget_facet_log` made to ignore `dropped` and always emit the "dropped" line — i.e. the pre-fix behaviour. `forget_facet_log_does_not_claim_a_drop_for_an_absent_key` failed with:

```
an absent key must not be logged as a state change — this is the #6108 defect, where the
claim was made before the row was read:
["learning.forget_facet: key=style/never_observed state=dropped user_state=forgotten"]
```

`forget_facet_log_claims_a_drop_only_when_a_row_was_written` correctly still passed under that fault — the fault makes the log always claim a drop, so the true case is unaffected. Fault reverted; `git status` is clean of it.

## Related

- Closes: Closes #6108
- Follow-up PR(s)/TODOs: none.

**Note for whoever merges:** #6104 also edits `handle_forget_facet` in this file (it adds a class-taxonomy check at the top of the function; this changes the log construction lower down). The two are complementary but will need a trivial rebase for whichever lands second — happy to rebase this one onto #6104.

---

### Linear Issue

- Key: N/A
- URL: N/A

### Commit & Branch

- Branch: `fix/6108-forget-facet-honest-log`
- Commit SHA: `0551356d2d9fe36c35958502651e71a694d313ba`

### Validation Run

- [x] `pnpm --filter openhuman-app format:check` — N/A: no frontend files changed.
- [x] `pnpm typecheck` — N/A: no TypeScript changed.
- [x] Focused tests: `cargo test --lib --features "$(bash scripts/ci/product-features.sh)" -- forget_facet_log` → 2 passed.
- [x] Rust fmt/check (if changed): `cargo fmt` clean, no changes produced.
- [x] Tauri fmt/check (if changed): N/A: no Tauri/shell code changed.

### Validation Blocked

- `command:` N/A
- `error:` N/A
- `impact:` N/A

### Behavior Changes

- Intended behavior change: the `forget_facet` log line no longer claims a state change when no row was written.
- User-visible effect: none in the RPC response; only in logs and support transcripts, which now distinguish "dropped" from "was not there".

### Parity Contract

- Legacy behavior preserved: yes — the response payload (`{"facet": …}` / `{"facet": null}`) and the success/error contract are byte-identical.
- Guard/fallback/dispatch parity checks: `tools.rs`'s `LearningForgetFacetTool` verified to already return `Value::Null` for an absent key, so the RPC and agent-tool surfaces agree.

### Duplicate / Superseded PR Handling

- Duplicate PR(s): none.
- Canonical PR: this one.
- Resolution: N/A

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved facet-forgetting behavior so logs accurately indicate whether a facet was removed or was not present.
  - Repeated attempts to forget an absent facet now remain non-mutating and clearly report that no change occurred.
  - Forgetting an existing facet records its dropped or forgotten state only when the facet is actually removed.

- **Tests**
  - Added regression coverage to verify accurate logging for both existing and absent facets.

### CODE REVIEW

**Comment by M3gA-Mind** [comment]
Both findings are the same finding, and it is a false positive. `forget_facet_log` **is** defined, in the same module the test file is compiled into.

The learning schemas module is assembled by `include!`, which the rule does not appear to model:

```rust
// src/openhuman/agent/learning/schemas.rs
#[cfg(test)]
#[path = "schemas_tests.rs"]
mod tests;                          // ← the test module
include!("schemas_part_01.rs");
include!("schemas_part_02.rs");     // ← defines forget_facet_log at :241 (this PR)
```

`include!` splices both `_part_NN.rs` files into `schemas.rs`'s own module, so `fn forget_facet_log` and `mod tests` are siblings in one module and the test file's `use super::*;` (line 1) resolves it. There is nothing to import — an explicit import would not even compile, because there is no `schemas_part_02` module path to import *from*.

The check's own output says why it reached the wrong conclusion:

> _The code index is behind this pull request (indexed at `d1094a1dd3e3`), so retrieved context may be out of date._

Verification, since "it compiles" is the whole claim:

- `cargo test --lib --features "$(bash scripts/ci/product-features.sh)" -- forget_facet_log` → **2 passed**. A missing definition is a compile error, so the tests could not have run at all.
- The Rust lanes on this PR are green, including `Rust Quality (fmt, clippy)`.
- I also revert-checked these two tests: with `forget_facet_log` made to ignore its `dropped` argument (the pre-fix behaviour), `forget_facet_log_does_not_claim_a_drop_for_an_absent_key` fails with *"an absent key must not be logged as a state change — this is the #6108 defect…"*. That is only possible if the function resolves.

On the alternative suggested — "replace the tests with an integration test that captures the actual log output from the real `handle_forget_facet` path" — that is a reasonable shape in general, but here it would need a bound memory cache and a live `FacetCache`, which is exactly why the log construction was split into a pure function. The `RpcOutcome`'s log vector is not observable over the JSON-RPC surface when the outcome is built with `RpcOutcome::new(..)`, so an integration test could assert the payload but not the log line — and the log line is the entire defect in #6108.

No change made. Happy to be shown wrong if there is a build configuration where this file does not see `schemas_part_02.rs`.

File: src/openhuman/agent/learning/schemas_tests.rs [hunk 1]
```diff
@@ -192,3 +192,44 @@ async fn class_taking_handlers_keep_presence_check_before_value_check() {
...
+
+/// The line was previously built unconditionally, before the read that decides
+/// whether there is anything to drop. A typo'd key therefore produced a log
+/// asserting `state=dropped user_state=forgotten` for a row that was never
+/// touched. The RPC contract itself is unchanged and deliberately idempotent —
+/// `{"facet": null}`, no error — so the log is the only thing that can tell the
+/// two outcomes apart.
+#[test]
+fn forget_facet_log_claims_a_drop_only_when_a_row_was_written() {
+    let dropped = forget_facet_log("style/observed_key", true);
```
[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/openhuman/agent/learning/schemas_tests.rs:206]: False positive — `forget_facet_log` is defined, in the same module this test file compiles into.

`src/openhuman/agent/learning/schemas.rs` assembles the module with `include!`, which the rule does not model:

```rust
#[cfg(test)]
#[path = "schemas_tests.rs"]
mod tests;                       // ← this file
include!("schemas_part_01.rs");
include!("schemas_part_02.rs");  // ← defines forget_facet_log (this PR)
```

`include!` splices both `_part_NN.rs` files into `schemas.rs`'s own module, so `fn forget_facet_log` and `mod tests` are siblings there and this file's `use super::*;` (line 1) resolves it. An explicit `use` would not compile — there is no `schemas_part_02` module path to import from, because it is not a module.

Evidence, since "it does not compile" is the whole claim:

- `cargo test --lib --features "$(bash scripts/ci/product-features.sh)" -- forget_facet_log rejects_unknown_class` → **9 passed, 0 failed** (my 2 plus the 7 class-validation tests #6104 added to this same file).
- Revert-checked: with `forget_facet_log` made to ignore its `dropped` argument, `forget_facet_log_does_not_claim_a_drop_for_an_absent_key` fails with *"an absent key must not be logged as a state change — this is the #6108 defect…"*. Only possible if the symbol resolves.

Worth noting this file is a known trap for symbol-resolution rules: openhuman#5856/#5857 split 205 source files into `_part_NN.rs` fragments pulled in by `include!`, so "not defined in the diff and not imported" will read as undefined for any of them.

No change made.

[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/openhuman/agent/learning/schemas_tests.rs:206]: False positive — `forget_facet_log` is defined, in the same module this test file compiles into.

`src/openhuman/agent/learning/schemas.rs` assembles the module with `include!`, which the rule does not model:

```rust
#[cfg(test)]
#[path = "schemas_tests.rs"]
mod tests;                       // ← this file
include!("schemas_part_01.rs");
include!("schemas_part_02.rs");  // ← defines forget_facet_log (this PR)
```

`include!` splices both `_part_NN.rs` files into `schemas.rs`'s own module, so `fn forget_facet_log` and `mod tests` are siblings there and this file's `use super::*;` (line 1) resolves it. An explicit `use` would not compile — there is no `schemas_part_02` module path to import from, because it is not a module.

Evidence, since "it does not compile" is the whole claim:

- `cargo test --lib --features "$(bash scripts/ci/product-features.sh)" -- forget_facet_log rejects_unknown_class` → **9 passed, 0 failed** (my 2 plus the 7 class-validation tests #6104 added to this same file).
- Revert-checked: with `forget_facet_log` made to ignore its `dropped` argument, `forget_facet_log_does_not_claim_a_drop_for_an_absent_key` fails with *"an absent key must not be logged as a state change — this is the #6108 defect…"*. Only possible if the symbol resolves.

Worth noting this file is a known trap for symbol-resolution rules: openhuman#5856/#5857 split 205 source files into `_part_NN.rs` fragments pulled in by `include!`, so "not defined in the diff and not imported" will read as undefined for any of them.

No change made.

**Comment by M3gA-Mind** [comment]
## Re: `tinysweeper/critique` — `Define or import forget_facet_log before using it`

This finding is a false positive. The helper **is** added by this PR, and the tests **do** compile.

**Where it is defined:** `src/openhuman/agent/learning/schemas_part_02.rs:264` —

```rust
fn forget_facet_log(full_key: &str, dropped: bool) -> Vec<String> {
```

That line is a `+` in this PR's own diff, in one of the two files the lane says it reviewed. It is
not pre-existing in `upstream/main`; this PR introduces both the helper and the tests that exercise
it.

**Why no `use` is needed.** `schemas.rs` assembles the module from fragments:

```rust
#[cfg(test)]
#[path = "schemas_tests.rs"]
mod tests;
include!("schemas_part_01.rs");
include!("schemas_part_02.rs");
```

`schemas_part_02.rs` is `include!`d directly into `schemas`, so `forget_facet_log` is an item **of
that module**, not of a separate one. `schemas_tests.rs` opens with `use super::*`, which brings it
into scope. A `use` naming another module would not compile, because there is no other module — the
`_part_NN.rs` split is textual inclusion, introduced repo-wide by #5856/#5857.

**Verified, not assumed.** Built and run at this head against the product feature set:

```
cargo test -p openhuman --lib --features "$(bash scripts/ci/product-features.sh)" \
  --no-default-features -- learning::schemas

test ...::forget_facet_log_claims_a_drop_only_when_a_row_was_written ... ok
test ...::forget_facet_log_does_not_claim_a_drop_for_an_absent_key ... ok
test result: ok. 20 passed; 0 failed
```

**Two things in the lane's own output already point at this.** It reports the item under
*"Fixed since the last review"* — twice — while still failing on it, and it notes *"The code index
is behind this pull request (indexed at `333782581086`), so retrieved context may be out of date."*
The sibling `tinysweeper/tests` lane reached the correct conclusion on the same diff: *"the earlier
finding about `forget_facet_log` being undefined is resolved by the introduction of that function."*

**Suggested rule improvement:** `import-resolution` appears to resolve symbols per-file. In this
repository a large number of modules are assembled from `_part_NN.rs` fragments via `include!`, so
a definition in a sibling fragment is in scope with no import. Treating `include!`d fragments as one
scope would remove a whole class of false positives here — this is the second time today a tool has
been misled by that split.

No change made; the finding does not hold.
